src/test/net_tests.cpp

```diff
@@ -771,4 +773,128 @@ BOOST_AUTO_TEST_CASE(PoissonNextSend)
     g_mock_deterministic_tests = false;
 }
 
+std::vector<NodeEvictionCandidate> GetRandomNodeEvictionCandidates(const int n_candidates, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates;
+    for (int id = 0; id < n_candidates; ++id) {
+        candidates.push_back({/* id */ id,
+                              /* nTimeConnected */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* nMinPingUsecTime */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* nLastBlockTime */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* nLastTXTime */ static_cast<int64_t>(random_context.randrange(100)),
+                              /* fRelevantServices */ random_context.randbool(),
+                              /* fRelayTxes */ random_context.randbool(),
+                              /* fBloomFilter */ random_context.randbool(),
+                              /* nKeyedNetGroup */ random_context.randrange(100),
+                              /* prefer_evict */ random_context.randbool(),
+                              /* m_is_local */ random_context.randbool()});
+    }
+    return candidates;
+}
+
+// Returns true if any of the node ids in node_ids are selected for eviction.
+bool IsEvicted(std::vector<NodeEvictionCandidate> candidates, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    Shuffle(candidates.begin(), candidates.end(), random_context);
+    const Optional<NodeId> evicted_node_id = SelectNodeToEvict(std::move(candidates));
+    if (!evicted_node_id) {
+        return false;
+    }
+    return std::find(node_ids.begin(), node_ids.end(), *evicted_node_id) != node_ids.end();
+}
+
+// Create number_of_nodes random nodes, apply setup function candidate_setup_fn,
+// apply eviction logic and then return true if any of the node ids in node_ids
+// are selected for eviction.
+bool IsEvicted(const int number_of_nodes, std::function<void(NodeEvictionCandidate&)> candidate_setup_fn, const std::vector<NodeId> node_ids, FastRandomContext& random_context)
+{
+    std::vector<NodeEvictionCandidate> candidates = GetRandomNodeEvictionCandidates(number_of_nodes, random_context);
+    for (size_t i = 0; i < candidates.size(); ++i) {
+        candidate_setup_fn(candidates[i]);
+    }
+    return IsEvicted(candidates, node_ids, random_context);
+}
+
+namespace {
+// The number of eviction candidates at (or above) which we are guaranteed that
+// the current eviction logic will find a node to evict no matter what eviction
+// candidates it is given. This constant may need to be adjusted if the eviction
+// logic changes.
+constexpr int GUARANTEED_EVICTION_AT_N_CANDIDATES_OR_ABOVE{29};
```

Comment:
Please add some justification for these numbers, even if it's a description of the simulation/test you ran to get them. I don't understand how this 29 is guaranteed and the 20 is guaranteed below. 

This reply makes it sound like it's *not* guaranteed?

https://github.com/bitcoin/bitcoin/pull/20477#discussion_r536421598

Reply:
Good discussion in PR review club about it today and @jonatack pointed out justification for the 20 is in `test/functional/p2p_eviction.py`.